File: binaryTree.js
const createNode = function (value) {
  return { value, left: null, right: null };
};

const insert = function (tree, value) {
  if (tree == null) return createNode(value);
  if (value < tree.value) tree.left = insert(tree.left, value)
  else tree.right = insert(tree.right, value);
  return tree;
};

const printInOrder = function (tree) {
  if (tree == null) return '';
  printInOrder(tree.left);
  process.stdout.write(`${tree.value} `);
  printInOrder(tree.right);
};

const printPreOrder = function (tree) {
  if (tree == null) return '';
  process.stdout.write(`${tree.value} `);
  printPreOrder(tree.left);
  printPreOrder(tree.right);
};

const printPostOrder = function (tree) {
  if (tree == null) return '';
  printPostOrder(tree.left);
  printPostOrder(tree.right);
  process.stdout.write(`${tree.value} `);
};

const search = function (tree, value) {
  if (tree == null) return false;
  if (tree.value == value) return true;
  if (value < tree.value) return search(tree.left, value);
  return search(tree.right, value);
};

const getParent = function (tree, value) {
  const isCorrectNode =
    (tree.left && tree.left.value == value) ||
    (tree.right && tree.right.value == value);
  if (isCorrectNode) return tree;
  if (value < tree.value) {
    return getParent(tree.left, value);
  } else {
    return getParent(tree.right, value);
  }
};

const getMaxNode = function (tree) {
  if (tree.right == null) return tree;
  return getMaxNode(tree.right);
};

const getMinNode = function (tree) {
  if (tree.left == null) return tree;
  return getMinNode(tree.left);
};

const getCompatibleNode = function (tree) {
  if (tree.left != null) return getMaxNode(tree.left);
  if (tree.right != null) return getMinNode(tree.right);
  return tree;
};

const deleteLeafNode = function (parent, node) {
  if (parent.right && parent.right.value == node.value) parent.right = null;
  if (parent.left && parent.left.value == node.value) parent.left = null;
};

const deleteNode = function (tree, node_to_delete) {
  const compatibleNode = getCompatibleNode(node_to_delete);
  const parent = getParent(tree, compatibleNode.value);
  node_to_delete.value = compatibleNode.value;
  if (compatibleNode.right == null && compatibleNode.left == null) {
    deleteLeafNode(parent, compatibleNode);
    return tree;
  }
  return deleteNode(tree, compatibleNode);
};

const rotateRight = function (node) {
  if (node == null || node.left == null) return node;
  const node_left = node.left;
  node.left = node_left.right;
  node_left.right = node;
  return node_left;
};

const rotateLeft = function (node) {
  if (node == null || node.right == null) return node;
  const node_right = node.right;
  node.right = node_right.left;
  node_right.left = node;
  return node_right;
};

const depth = function (node) {
  if (node === null) return 0;
  const left_depth = depth(node.left);
  const right_depth = depth(node.right);
  return left_depth > right_depth ? left_depth + 1 : right_depth + 1;
};

const getBalanceFactor = function (node) {
  const left_depth = depth(node.left);
  const right_depth = depth(node.right);
  return right_depth - left_depth;
};

const isLeafNode = function (node) {
  console.log(node, '---')
  return node.right == null && node.left == null;
};

const balance = function (node) {
  if (node === null) return node;
  const balance_factor = getBalanceFactor(node);
  if (balance_factor >= -1 && balance_factor <= 1) {
    node.left = balance(node.left);
    node.right = balance(node.right);
    return node;
  }
  node = balance_factor < 0 ? rotateRight(node) : rotateLeft(node);
  return balance(node);
};

const printTree = function (tree) {
  console.log(JSON.stringify(tree), '\n');
};

const tree = [4, 3, 5, 2, 6].reduce(insert, null);

const balancedTree = balance(tree);
printTree(balancedTree);

File: binary_tree.c
#include <stdlib.h>
#include "binary_tree.h"

node_ptr create_node(Element e)
{
    node_ptr new_node = malloc(sizeof(node));
    new_node->element = e;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}

void insert_node(node_ptr root, Comparer compare, Element e)
{
    node_ptr *p_walker = &root;
    while (*p_walker != NULL)
    {
        if (compare(e, (*p_walker)->element) == Less)
            p_walker = &(*p_walker)->left;
        else
            p_walker = &(*p_walker)->right;
    }
    *p_walker = create_node(e);
}

void print_in_order(node_ptr root, Displayer display)
{
    if (root == NULL)
        return;
    print_in_order(root->left, display);
    display(root);
    print_in_order(root->right, display);
}
void print_pre_order(node_ptr root, Displayer display)
{
    if (root == NULL)
        return;
    display(root);
    print_pre_order(root->left, display);
    print_pre_order(root->right, display);
}
void print_post_order(node_ptr root, Displayer display)
{
    if (root == NULL)
        return;
    print_post_order(root->left, display);
    print_post_order(root->right, display);
    display(root);
}

Bool search(node_ptr root, Comparer compare, Element e)
{
    node_ptr *p_walker = &root;
    while (*p_walker != NULL)
    {
        p_walker = &(*p_walker)->right;
        Equality result = compare(e, (*p_walker)->element);
        if (result == Equal)
            return True;
        if (result == Less)
            p_walker = &(*p_walker)->left;
    }
    return False;
}

node_ptr get_parent(node_ptr root, Element e, Comparer compare)
{
    Bool left_child_check = root->left != NULL && root->left->element == e;
    Bool right_child_check = root->right != NULL && root->right->element == e;
    if (left_child_check || right_child_check)
        return root;
    if (compare(e, root->element) == Less)
        return get_parent(root->left, e, compare);
    return get_parent(root->right, e, compare);
}

node_ptr get_max_node(node_ptr root)
{
    if (root->right == NULL)
        return root;
    return get_max_node(root->right);
}

node_ptr get_min_node(node_ptr root)
{
    if (root->left == NULL)
        return root;
    return get_min_node(root->left);
}

node_ptr get_compatible_node(node_ptr root)
{
    if (root->left != NULL)
        return get_max_node(root->left);
    if (root->right != NULL)
        return get_min_node(root->right);
    return root;
}

void free_node(node_ptr node)
{
    free(node->element);
    free(node);
}

Bool is_leaf_node(node_ptr root)
{
    return root->right == NULL && root->left == NULL;
}

void delete_leaf_node(node_ptr parent, node_ptr to_be_deleted, Comparer compare)
{
    if (parent->left != NULL && parent->left == to_be_deleted)
        parent->left = NULL;
    if (parent->right != NULL && parent->right == to_be_deleted)
        parent->right = NULL;
    free_node(to_be_deleted);
}

node_ptr delete (node_ptr root, node_ptr node_to_delete, Comparer compare, Copier copy)
{
    node_ptr compatible_node = get_compatible_node(node_to_delete);
    node_ptr parent = get_parent(root, compatible_node->element, compare);
    free(node_to_delete->element);
    node_to_delete->element = copy(compatible_node->element);
    if (is_leaf_node(compatible_node))
    {
        delete_leaf_node(parent, compatible_node, compare);
        return root;
    }
    return delete (root, compatible_node, compare, copy);
}

node_ptr right_rotate(node_ptr node_to_rotate)
{
    if (node_to_rotate == NULL || node_to_rotate->left == NULL)
        return node_to_rotate;
    node_ptr node_to_rotate_left = node_to_rotate->left;
    node_to_rotate->left = node_to_rotate_left->right;
    node_to_rotate_left->right = node_to_rotate;
    return node_to_rotate_left;
}

node_ptr left_rotate(node_ptr node_to_rotate)
{
    if (node_to_rotate == NULL || node_to_rotate->right == NULL)
        return node_to_rotate;
    node_ptr node_to_rotate_right = node_to_rotate->right;
    node_to_rotate->right = node_to_rotate_right->left;
    node_to_rotate_right->left = node_to_rotate;
    return node_to_rotate_right;
}

void count_nodes(node_ptr root, int_ptr count)
{
    if (root == NULL)
        return;
    count_nodes(root->left, count);
    *count = *count + 1;
    count_nodes(root->right, count);
}

void store_nodes_in_order(node_ptr *array, node_ptr root, int *index)
{
    if (root == NULL)
        return;
    store_nodes_in_order(array, root->left, index);
    array[*index] = root;
    *index = *index + 1;
    store_nodes_in_order(array, root->right, index);
}

node_ptr *get_node_in_order(node_ptr root, int_ptr index)
{
    int_ptr node_count = malloc(sizeof(int_ptr));
    *node_count = 0;
    count_nodes(root, node_count);
    node_ptr *nodes = malloc(sizeof(node_ptr) * (*node_count));
    store_nodes_in_order(nodes, root, index);
    return nodes;
}

node_ptr perform_balance(node_ptr *nodes, int start_idx, int end_idx, Comparer compare)
{
    if (start_idx > end_idx)
        return NULL;
    int mid_idx = (int)(start_idx + end_idx) / 2;
    node_ptr root = create_node(nodes[mid_idx]->element);
    root->left = perform_balance(nodes, start_idx, mid_idx - 1, compare);
    root->right = perform_balance(nodes, mid_idx + 1, end_idx, compare);
    return root;
}

node_ptr balance(node_ptr root, Comparer compare)
{
    int_ptr index = malloc(sizeof(int));
    *index = 0;
    node_ptr *nodes = get_node_in_order(root, index);
    return perform_balance(nodes, 0, *index - 1, compare);
}

int max(int x, int y)
{
    if (x > y)
        return x;
    else
        return y;
}

int get_tree_depth(node_ptr root)
{
    if (root == NULL)
        return 0;
    int left_tree_depth = get_tree_depth(root->left);
    int right_tree_depth = get_tree_depth(root->right);
    return max(left_tree_depth, right_tree_depth) + 1;
}

Bool get_balance_factor(node_ptr root)
{
    int left_tree_depth = get_tree_depth(root->left);
    int right_tree_depth = get_tree_depth(root->right);
    return right_tree_depth - left_tree_depth;
}

node_ptr balance_with_rotation(node_ptr root, Comparer compare)
{
    if (root == NULL)
        return root;
    int balance_factor = get_balance_factor(root);
    if (balance_factor >= -1 && balance_factor <= 1)
    {
        root->left = balance_with_rotation(root->left, compare);
        root->right = balance_with_rotation(root->right, compare);
        return root;
    }
    root = balance_factor < 0 ? right_rotate(root) : left_rotate(root);
    return balance_with_rotation(root, compare);
}